internal/app/go-filecoin/internal/submodule/storage_miner_node.go

```diff
@@ -0,0 +1,508 @@
+// given height.
+type ChainSampler func(ctx context.Context, sampleHeight *types.BlockHeight) ([]byte, error)
+
+type heightThresholdListener struct {
+	target    uint64
+	targetHit bool
+
+	seedCh    chan storage.SealSeed
+	errCh     chan error
+	invalidCh chan struct{}
+	doneCh    chan struct{}
+}
+
+// handle a chain update by sending appropriate status messages back to the channels.
+// newChain is all the tipsets that are new since the last head update.
+// Normally, this will be a single tipset, but in the case of a re-org it will contain
+// All the common ancestors of the new tipset to the greatest common ancestor.
+// Returns false if this handler is no longer valid
+func (l heightThresholdListener) handle(ctx context.Context, chain []block.TipSet, sampler ChainSampler) (bool, error) {
+	if len(chain) < 1 {
+		return true, nil
+	}
+
+	h, err := chain[0].Height()
+	if err != nil {
+		return true, err
+	}
+
+	// check if we've hit finality and should stop listening
+	if h >= l.target+consensus.FinalityEpochs {
+		return false, nil
+	}
+
+	lcaHeight, err := chain[len(chain)-1].Height()
+	if err != nil {
+		return true, err
+	}
+
+	// if we have already seen a target tipset
+	if l.targetHit {
+		// if we've completely reverted
+		if h < l.target {
+			l.invalidCh <- struct{}{}
+			l.targetHit = false
+			// if we've re-orged to a point before the target
+		} else if lcaHeight < l.target {
+			l.invalidCh <- struct{}{}
+			err := l.sendRandomness(ctx, chain, sampler)
+			if err != nil {
+				return true, err
+			}
+		}
+		return true, nil
+	}
+
+	// otherwise send randomness if we've hit the height
+	if h >= l.target {
+		l.targetHit = true
+		err := l.sendRandomness(ctx, chain, sampler)
+		if err != nil {
+			return true, err
+		}
+	}
+	return true, nil
+}
+
+func (l heightThresholdListener) sendRandomness(ctx context.Context, chain []block.TipSet, sampler ChainSampler) error {
+	// assume chain not empty and first tipset height greater than target
+	firstTargetTipset := chain[0]
+	for _, ts := range chain {
+		h, err := ts.Height()
+		if err != nil {
+			return err
+		}
+
+		if h < l.target {
+			break
+		}
+		firstTargetTipset = ts
+	}
+
+	tsHeight, err := firstTargetTipset.Height()
+	if err != nil {
+		return err
+	}
+
+	randomness, err := sampler(ctx, types.NewBlockHeight(tsHeight))
+	if err != nil {
+		return err
+	}
+
+	l.seedCh <- storage.SealSeed{
+		BlockHeight: tsHeight,
+		TicketBytes: randomness,
+	}
+	return nil
+}
+
+// StorageMinerNodeAdapter is a struct which satisfies the go-storage-miner
+// needs of "the node," e.g. interacting with the blockchain, persisting sector
+// states to disk, and so forth.
+type StorageMinerNodeAdapter struct {
+	minerAddr  address.Address
+	workerAddr address.Address
+
+	newListener     chan heightThresholdListener
+	heightListeners []heightThresholdListener
+	listenerDone    chan struct{}
+
+	chain     *ChainSubmodule
```

Comment:
> This should be implemented with direct access to the dependencies: chain reader instead of chain submodule, messaging outbox instead of messaging submodule etc.

@laser will handle this one

Reply:
done